cdap-examples/WikipediaPipeline/src/main/java/co/cask/cdap/examples/wikipedia/WikiContentValidatorAndNormalizer.java

```diff
@@ -0,0 +1,192 @@
+
+  @Override
+  protected void configure() {
+    setName(NAME);
+    setDescription("A MapReduce program that dumps page titles to a dataset.");
+    setMapperResources(new Resources(512));
+    setInputDataset(WikipediaPipelineApp.RAW_WIKIPEDIA_DATASET);
+    setOutputDataset(WikipediaPipelineApp.NORMALIZED_WIKIPEDIA_DATASET);
+  }
+
+  @Override
+  public void beforeSubmit(MapReduceContext context) throws Exception {
+    Job job = context.getHadoopJob();
+    job.setMapperClass(FilterNormalizerMapper.class);
+    job.setNumReduceTasks(0);
+  }
+
+  @Override
+  public void onFinish(boolean succeeded, MapReduceContext context) throws Exception {
+    WorkflowToken workflowToken = context.getWorkflowToken();
+    if (workflowToken != null) {
+      workflowToken.put("result", Value.of(succeeded));
+    }
+  }
+
+  /**
+   * Mapper that:
+   * - Filters records that are null or empty or cannot be parsed as JSON.
+   * - Removes meta fields from the raw Wikipedia JSON blobs.
+   * - Normalizes data formatted as Wikitext to plain text.
+   */
+  public static class FilterNormalizerMapper extends Mapper<byte [], byte [], byte [], byte []> {
+    private static final Logger LOG = LoggerFactory.getLogger(FilterNormalizerMapper.class);
+    private static final Gson GSON = new Gson();
+
+    @Override
+    protected void map(byte[] key, byte[] value, Context context) throws IOException, InterruptedException {
+      if (key == null) {
+        LOG.debug("Found null key. Skipping record.");
+        return;
+      }
+      if (key.length == 0) {
+        LOG.debug("Found empty key. Skipping record.");
+        return;
+      }
+      if (value == null) {
+        LOG.debug("Found null value. Skipping record.");
+        return;
+      }
+      if (value.length == 0) {
+        LOG.debug("Found null value. Skipping record.");
+        return;
+      }
+      WikiTitleAndText titleAndText;
+      try {
+        titleAndText = parse(value);
+      } catch (JsonSyntaxException e) {
+        LOG.debug("Malformed JSON found as value. Wikipedia may not have an entry for the page '{}'",
+                  Bytes.toString(key));
+        return;
+      }
+      String plainText;
+      try {
+        plainText = toPlainText(titleAndText);
+      } catch (EngineException | LinkTargetException e) {
+        LOG.debug("Error while parsing wikitext for '{}'. Skipping record.", Bytes.toString(key));
+        return;
+      }
+      context.write(key, Bytes.toBytes(plainText));
+      context.getCounter("custom", "num.records").increment(1);
+    }
+
+    private WikiTitleAndText parse(byte[] rawWikiData) {
+      WikiContents json = GSON.fromJson(Bytes.toString(rawWikiData), WikiContents.class);
+      Map<String, WikiContents.Query.Page> pages = json.query.pages;
+      // there is only one entry in this map, with the key as the page id
+      String key = pages.keySet().iterator().next();
+      WikiContents.Query.Page page = pages.get(key);
```

Comment:
you could rather do ``pages.entrySet().iterator().next().getValue()`` and remove the previous line to get key.